File: engine/Physics.js
/**
 * Physics.js
 * Top-Down 2.5D Physics
 * X, Y: Ground Plane
 * Z: Height (Jumping)
 */
export class Physics {
    constructor() {
        this.gravity = 0.9; // Higher gravity for snappy hops
        this.friction = 0.85;
        // Boundaries
        this.minY = 200; // Horizon line
        this.maxY = window.innerHeight - 50;
    }

    applyGravity(entity) {
        // Gravity affects Z axis now
        if (entity.z > 0 || entity.vz > 0) {
            entity.vz -= this.gravity;
            entity.isGrounded = false;
        }
    }

    applyPhysics(entity) {
        // Ground Movement (X/Y)
        entity.x += entity.vx;
        entity.y += entity.vy;
        
        // Height Movement (Z)
        entity.z += entity.vz;

        // Friction
        entity.vx *= this.friction;
        entity.vy *= this.friction;
        
        // Floor Collision (Z-axis)
        if (entity.z < 0) {
            entity.z = 0;
            entity.vz = 0;
            entity.isGrounded = true;
        }
        
        // Screen Boundaries (X/Y)
        if (entity.x < 0) entity.x = 0;
        if (entity.y < this.minY) entity.y = this.minY;
        if (entity.y > this.maxY) entity.y = this.maxY;
    }
    
    checkCollision(rect1, rect2) {
        // 2.5D Collision: Check Footprint (X/Y) AND Height (Z) overlap?
        // For now, we mainly care about ground footprint overlap
        return (
            rect1.x < rect2.x + rect2.width &&
            rect1.x + rect1.width > rect2.x &&
            rect1.y < rect2.y + rect2.height &&
            rect1.y + rect1.height > rect2.y
        );
    }
}

File: physics.py
# physics.py
from settings import *

class Physics:
    def __init__(self):
        self.gravity = GRAVITY
        self.friction = FRICTION
        self.min_y = GROUND_HORIZON
        self.max_y = WORLD_HEIGHT - 50

    def apply_gravity(self, entity):
        # Gravity affects Z axis (Height)
        if entity.z > 0 or entity.vz > 0:
            entity.vz -= self.gravity
            entity.is_grounded = False

    def apply_physics(self, entity):
        # Ground Movement (X/Y)
        entity.pos.x += entity.vel.x
        entity.pos.y += entity.vel.y
        
        # Height Movement (Z)
        entity.z += entity.vz
        
        # Friction
        entity.vel.x *= self.friction
        entity.vel.y *= self.friction
        
        # Floor Collision (Z-axis)
        if entity.z < 0:
            entity.z = 0
            entity.vz = 0
            entity.is_grounded = True
            
        # Screen Boundaries
        if entity.pos.x < 0: entity.pos.x = 0
        if entity.pos.x > WORLD_WIDTH - entity.rect.width: entity.pos.x = WORLD_WIDTH - entity.rect.width
        
        if entity.pos.y < self.min_y: entity.pos.y = self.min_y
        if entity.pos.y > WORLD_HEIGHT - entity.rect.height: entity.pos.y = WORLD_HEIGHT - entity.rect.height
        
        # Update Rect for drawing (Projected position)
        # We draw at (x, y - z)
        entity.rect.x = int(entity.pos.x)
        entity.rect.y = int(entity.pos.y - entity.z)
